library/src/main/java/MapUtils.java

```diff
@@ -0,0 +1,50 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Stack;
+
+/** Provides static functions for some operations on Maps */
+public class MapUtils {
+
+  /**
+   * Adds all root to leaf node keypaths in {@code map} to {@code keypaths} HashSet
+   *
+   * @param map the map which we will find all the keypaths for
+   * @param keypath the path of keys taken to get to a leaf node in the map
+   * @param keypaths a HashSet of Strings which this function will add all the keypaths of leaf
+   *     nodes to during the traversal
+   */
+  @SuppressWarnings("unchecked")
```

Comment:
From the style guide:

8.4.2 How to handle a warning
Before doing anything, make sure you understand exactly what the warning is telling you. Unchecked warnings in particular are infamously difficult to understand and frequently indicate a real problem. If you're not positive why a warning is appearing, ask for help (for example on the java‑users@google.com list).

Once you understand the warning, attempt the following solutions in order:

First, fix or work around.
Make a strong attempt to actually address the warning, or find another way to accomplish the task that avoids the situation entirely.

Otherwise, determine if it's a false alarm.
If you are convinced that the code is actually safe and correct, add a comment to convince the reader of this fact and apply the @SuppressWarnings annotation. Here are numerous examples of simple comments you can use in various situations.
Note: unless you have a convincing argument (and comment) that the warning is invalid (not just an argument that fixing it is infeasible or that you don't have time), do not use @SuppressWarnings; instead continue to...

If all else fails, leave a TODO comment.
This is a last resort. Refer to Section 4.8.6.2, TODO comments.
If you do this, do not suppress the warning. The warning should be visible until it can be taken care of properly.

Reply:
`SuppressWarnings` was used here and in a few other places in the code. When deserializing the YAML file, it becomes a `LinkedHashMap<String, Object>` which is a `Map<String, Object>`. The value of `map` could be another `Map`, or other stuff i.e. `List`, `String`, `Integer`, `Boolean`. This piece of recursive code assumes that the provided `Map` fits this criteria, namely that if the value is a `Map`, then it will always be a `Map<String, Object>`. Usage of this function is internal in the library where we can guarantee that the `map` parameter provided is `Map<String, Object>` where if the Object value is a `Map`, and it passes the `(value instanceof Map) condition, then it must be `Map<String, Object>`.

Alternatives tried/considered:

- Here, it may be possible to use Map<?,?> but it doesn't necessarily seem like a better way to go about it, and we might not be able to remove other unchecked casts in this way.
- If we create several custom recursive classes like `Node`, `IntegerNode`, `StringNode`... to represent the spec tree, then we may not have the unchecked issue right here, but we would have to transform the Map<String, Object> into this new custom class which will need to SuppressWarnings anyway. Then we will also need to convert this custom node class back to YAML file. For this reason and after discussion with Mike Yara it seemed like it wouldn't be worth it to do it this way either. SuppressWarnings seems safe here because we know what type of data will be coming into these functions and the runtime check should not fail.

Please let me know if you have any further feedback about this! otherwise I will add the comment and commit that.

Reply:
Is it possible to share this style guide with me? I would like to take a look at the "Here are numerous examples of simple comments you can use in various situations"